### PR TITLE
Script to create new projects

### PR DESCRIPTION
**Pull request by Ricocotam**
From issue #17  I created a minimal example of what I had in mind. This is a minimal example and some questions need to be answered.

### Path of the new project
We need to create a folder for the project. I propose to use 2 different name : project_path and project_name. We then could have an architecture like :

```
+ project_path
    + project_name
        - __init__.py
        + models
            + networks
                - __init__.py
                - factory.py
                - mynetwork.py
    + logs
```
(`+` denotes a folder, `-` a file)

This is an open suggestion. I choose this implementation so logs' folder is also created and is with the rest of the code.

### Network as minimal example
I did a minimal example so we can discuss implementation details. Creating files and folders is done with `pathlib.Path` to use the `/` operation to join paths. The real question is : how to we fill created files with basic content ? I choose to have raw text written in the `new.py` I am Pull Requesting but I don't think that's the better option as this is not easily maintainable and it leads to a really long `new.py` file. My suggestion is to have a basic file we can copy to the project created. This requires to store quite a lot of files in the lib only used to be copied so it has some drawbacks too.

I am also working on a complete project creation with options to create any relevant directory but I didn't want to pollute the discussion with useless lines of code

### CODE REVIEW

**Comment by Cadene** [comment]
@Ricocotam @MicaelCarvalho 

Much needed!

However, we need a way to avoid putting executable code into a string such as [this](https://github.com/Cadene/bootstrap.pytorch/pull/19/files#diff-62d640f6d51184c92678ea3bf632ba5dR26-R48)

I would suggest creating template python files in the bootstrap directory, copying them and even replacing some elements in it (maybe???).

For instance, instead of [this](https://github.com/Cadene/bootstrap.pytorch/pull/19/files#diff-62d640f6d51184c92678ea3bf632ba5dR26-R48), we would have the following file in `bootstrap/template/project/models/networks/network.py`:

```python
import torch.nn as nn
class MyNetwork(nn.Module):
    def __init__(self, *args, **kwargs):
        super(MyNetwork, self).__init__()
        # Assign args
    def forward(self, x):
        # x is a dictionnary given by Dataset class
        pred = self.net(x)
        return pred  # This is a tensor (or several tensors)
```

**Comment by Ricocotam** [comment]
Alright, so I did a sample of this on the latest commit

File: bootstrap/new.py [hunk 1]
```diff
@@ -1,64 +1,88 @@
+import os
 from pathlib import Path
 from argparse import ArgumentParser
 
 
+file_dir = Path(__file__).parent
+
+
 parser = ArgumentParser()
-parser.add_argument("project_path", action="store", type=str, help="Path to new project")
```
Comment (Cadene) [inline bootstrap/new.py]: We might want to keep `project_path`.
What do you think @MicaelCarvalho ?

File: bootstrap/new.py [hunk 2]
```diff
@@ -1,64 +1,88 @@
...
+        "models/criterions",
+        "models/metrics",
+    ]
+
+    for directory in directories:
+        print(f"Creating {directory} folder and associated files")
+        new_dir = path / directory
+        if directory != "models":
+            new_dir.mkdir()
+        files = get_files(new_dir)
```
Comment (Cadene) [inline bootstrap/new.py]: As specified [here](https://github.com/Cadene/bootstrap.pytorch/pull/19#issuecomment-605525218), we would like to follow this path pattern `bootstrap/templates/default/project/datasets/dataset.py` instead of your current `bootstrap/template/datasets/template_dataset.py`.

File: bootstrap/template/models/metrics/factory.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from bootstrap.lib.options import Options
+
+from .mymetric import {PROJECT_NAME}Metric
+
+
+def factory(engine=None, mode="train"):
+    opt = Options()['model.metric']
+
+    if opt['name'] == '{PROJECT_NAME_LOWER}metric':
```
Comment (Cadene) [inline bootstrap/template/models/metrics/factory.py]: `'metric'` can be removed here

File: bootstrap/template/models/metrics/factory.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+        metric = {PROJECT_NAME_LOWER}metric()
```
Comment (Cadene) [inline bootstrap/template/models/metrics/factory.py]: Metric

**Comment by Cadene** [comment]
Thanks! @Ricocotam 
I added some comments

We should probably add a test to verify that everything can work.
It implies to have some default return values (`torch.tensor(0)`) instead of `raise NotImplementedError` or `pass`.
What do you think? @Ricocotam @MicaelCarvalho

File: bootstrap/template/models/metrics/factory.py (hunk 3 above)
Comment (Ricocotam) [reply bootstrap/template/models/metrics/factory.py:9]: Not sure to understand what you mean ?

**Comment by Ricocotam** [comment]
I'll fix your different comments soon. Anything to add ?

File: bootstrap/template/models/metrics/factory.py (hunk 3 above)
Comment (Cadene) [reply bootstrap/template/models/metrics/factory.py:9]:
```
def factory(engine=None, mode="train"):
    opt = Options()['model.metric']

    if opt['name'] == '{PROJECT_NAME_LOWER}':
```

**Comment by Ricocotam** [comment]
So we need some tests now

The only thing I didn't do is to name custom dataset after the project name, I kept `mydataset.py` : `myexample.bootstrap.pytorch/myexample/datasets/myexample.py`-> `myexample.bootstrap.pytorch/myexample/datasets/mydataset.py`

I can write some tests (we just have to check files have the right content and that they are created) but I know how it works already. May be, to ensure code is both clear and that agree on what/how it does it, someone else should right the tests

**Comment by Ricocotam** [comment]
Solved in #35

**Comment by MicaelCarvalho** [comment]
This PR broke `master` and all tests... Please be more careful and make sure tests pass before merging ;)
